src/mpid/ch4/include/mpidpre.h

```diff
@@ -227,45 +227,57 @@ struct MPIDI_CH4U_win_lock {
     uint16_t type;
 };
 
-struct MPIDI_CH4U_win_queue {
+typedef struct MPIDI_CH4U_win_lock_recvd {
     struct MPIDI_CH4U_win_lock *head;
     struct MPIDI_CH4U_win_lock *tail;
-};
+    int type;                   /* current lock's type */
+    unsigned count;             /* count of granted locks (not received) */
+} MPIDI_CH4U_win_lock_recvd_t;
 
-typedef struct MPIDI_CH4U_win_lock_info {
-    unsigned peer;
-    int lock_type;
-    struct MPIR_Win *win;
-    volatile unsigned done;
-} MPIDI_CH4U_win_lock_info;
+typedef struct MPIDI_CH4U_win_target_sync_lock {
+    volatile unsigned locked;   /* locked == 0 or 1 */
```

Comment:
I agree your point that as long as the `pthread` lock functions are used, the compiler will not optimize the code by register caching it, thus we do not need `volatile` per variable. 

However, if other lock mechanism is used in future, such as Halim's `izem`, which can be fully inlined with zero function call, then the above assumption cannot work. There is actually a plan to add the configure option to allow user to specify the thread CS implementation. 

All I am saying is, the lock theory itself does not guarantee no register caching, it is implementation dependent.  `volatile` is the only precise way to avoid  register caching, which solves the specific problem of the RMA variables.